contracts/strategy/LinearYieldDistributionStrategy.sol

```diff
@@ -0,0 +1,178 @@
+     */
+    event StrategyNewYieldDistributionCycle(
+        uint256 startTimestamp,
+        uint256 duration,
+        uint256 distributionAmount,
+        uint256 startAmount
+    );
+
+    error StrategyInvalidYieldDistributionCycleDuration();
+
+    uint256 public constant MIN_CYCLE_DURATION = 1 days;
+    uint256 public constant MAX_CYCLE_DURATION = 21 days;
+
+    // The length of a yield distribution cycle
+    uint256 public cycleDuration;
+    // The amount being distributed in the yield distribution cycle
+    uint256 public cycleDistributionAmount;
+    // The timestamp of the start of the yield distribution cycle
+    uint256 public cycleStartTimestamp;
+    // The sum of deposits and distributed yield at the cycle start
+    uint256 public cycleStartAmount;
+    // The sum of deposits and distributed yield at the cycle end
+    uint256 public cycleEndAmount;
+
+    constructor(uint256 _yieldCycleDuration) {
+        _setYieldDistributionCycleDuration(_yieldCycleDuration);
+    }
+
+    /**
+     * Sets the duration of the yield distribution cycle.
+     */
+    function setYieldDistributionCycleDuration(uint256 _newDuration)
+        public
+        onlySettings
+    {
+        _setYieldDistributionCycleDuration(_newDuration);
+    }
+
+    /// @inheritdoc IStrategy
+    function investedAssets()
+        public
+        view
+        virtual
+        override(IStrategy)
+        returns (uint256)
+    {
+        uint256 totalInvestedAssets = _totalInvestedAssets();
+
+        // if there's no yield being distributed
+        if ((cycleStartTimestamp == 0) || (cycleDistributionAmount == 0))
+            return
+                cycleStartAmount < totalInvestedAssets
+                    ? cycleStartAmount
+                    : totalInvestedAssets;
+
+        uint256 cycleCurrentAmount = _calcCurrentlyDistributedYieldAmount();
+
+        // if there's less funds, return the real funds
+        if (totalInvestedAssets < cycleCurrentAmount)
+            return totalInvestedAssets;
+
+        return cycleCurrentAmount;
+    }
+
+    /**
+     * Updates the yield distribution cycle or starts a new one if previous has finished. This function can be called when
+     * there's a new yield generated or when there's a loss in the strategy.
+     *
+     * @notice It can be called by anyone because there's not harm from it, and it
+     * makes the system less reliant on the backend.
+     */
+    function updateYieldDistributionCycle() public {
+        uint256 totalInvestedAssets = _totalInvestedAssets();
+
+        // there is no new yield generated to distribute since cycle started
+        if (cycleEndAmount == totalInvestedAssets) return;
+
+        // if there was yield being distributed
+        if (cycleDistributionAmount > 0) {
+            cycleStartAmount = _calcCurrentlyDistributedYieldAmount();
+        }
+
+        // if funds were lost
```

Comment:
Yes, that's what I was talking about and it's actually true because when a deposit/withdrawal happens, both `cycleEndAmount` and `cycleStartAmount` are updated in appropriate handler functions.

Reply:
👍 makes sense